specs/agents/tracing-instrumentation-azure-functions.md

```diff
@@ -0,0 +1,103 @@
+
+| `faas.execution` | `203621a2-62f...` | The unique invocation id of the function. | *generic context* |
+| `faas.coldstart` | `true`/`false` | A boolean value indicating whether this function invocation was a cold start or not. | [see section below](deriving-cold-starts)
+
+### Metadata
+
+Generic cloud metadata fetching SHOULD be disabled when the agent is running in an Azure Functions app. It will not
+yield any useful data and only incur additional cold start overhead due to the non-existent HTTP metadata endpoints.
+
+Agents SHOULD rather:
+
+* Check for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable that indicates that the agent is running
+in an Azure Function application.
+* Perform specific cloud metadata detection for Azure Functions (see [Metadata](./metadata.md)).
+
+In addition to those described in [Metadata](./metadata.md), following metadata fields are relevant for Azure Functions:
+
+| Field | Value | Description | Source |
+| - | - | - | - |
+| `service.framework.name` | `Azure Functions` | Constant value for the framework name. | |
+| `service.runtime.name`| e.g. `dotnet-isolated` | The language worker runtime (see [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_runtime)). | `FUNCTIONS_WORKER_RUNTIME` |
+
+### Deriving cold starts
+
+A cold start occurs if the Azure Functions runtime needs to be initialized in order to handle the first function execution.
+Since an Azure Function app can provide multiple function entry points, those may run concurrently.
+Hence, cold start detection must happen in a thread-safe manner.
+
+**Note:** The Azure Functions [Premium plan](https://learn.microsoft.com/en-us/azure/azure-functions/functions-scale)
+guarantees pre-warmed workers (no cold starts). Nonetheless, reporting the first function execution as such might still make
+sense since it could come with additional processing due to function-specific initialization overhead or JIT costs.  
+Alternatively, agents could try to detect their hosting plan and not report cold starts for Premium plans at all.
+
+### Disabled Functionality
+
+Agents SHOULD detect whether they are running in an Azure Functions environment by testing
+for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable.
+
+The following agent functionalities SHOULD to be turned off when tracing Azure Functions until decided otherwise:
+
+* **Metrics collection:** this includes all kind of metrics: system, process and breakdown metrics and is equivalent to
+setting `ELASTIC_APM_METRICS_INTERVAL = 0`
+* **Remote configuration:** equivalent to setting `ELASTIC_APM_CENTRAL_CONFIG = false`
+* **Cloud metadata discovery:**
+  * Equivalent to setting `ELASTIC_APM_CLOUD_PROVIDER = none`
```

Comment:
I think this part clashes with the general idea of the `ELASTIC_APM_CLOUD_PROVIDER` config.

The point of that config is that the user just puts their cloud provider name there (in this case `azure`) and from that point the agent will only try to collect metadata for that provider.

This is maybe implementation detail, but if I remember correctly, setting this to `none` will also disable `AzureFunctionsMetadataProvider` in the .NET agent currently - meaning setting it to `none` will in general disable all cloud metadata collection - and this, in the current code includes the new one specifically for Azure Functions suggested in the line below. And this behaviour reflects the way this config is intended to work.

I'd expect other agents to work similarly.

The point of the spec here is clear (only use the new Azure Functions specific metadata collection suggested in the next line), I'm just not sure the `ELASTIC_APM_CLOUD_PROVIDER = none` reference is ideal.

Reply:
Could you take another look here @gregkalapos? Thanks

Reply:
I guess I understand the intention and it totally makes sense to me: let's not try to query AWS, Google, and in Azure non Azure Functions related endpoints.

> or - that's what I had in mind - turn off the generic mechanism and just run the Azure Functions specific code since we know we are running as such due the presence of certain environment variables.

Yes, I also think this is the way to go, but still I'd not make this equivalent to `ELASTIC_APM_CLOUD_PROVIDER = none`, because in https://github.com/elastic/apm/pull/712/ that'll be listed as any other metadata collection. So setting it to `none` would make me think that none of the metadata collection listed in that list will run.

This is what [the spec says](https://github.com/elastic/apm/blob/main/specs/agents/metadata.md#cloud-provider-metadata):
> This metadata collection is controlled by a configuration value, CLOUD_PROVIDER. ... 
If set to none, no cloud metadata will be generated.

And that's not the case here, right? We still do generate cloud metadata - or don't we?

I just looked at the .NET code and this part is not yet implemented there. 

Would it make sense for quickly hop on zoom and discuss? I feel we agree on what we want here, but still this `ELASTIC_APM_CLOUD_PROVIDER = none` part sounds a bit strange to me when I see the rest of the spec.

Reply:
Yes, I think we agree on the general idea 👍 .

I'm not married to the exact wording here.
In fact, I took this from the [AWS Lambda spec](https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-aws-lambda.md). So we might need to reconsider the wording around `ELASTIC_APM_CLOUD_PROVIDER = none` there as well, I guess.

> I just looked at the .NET code and this part is not yet implemented there.
That's right.

Reply:
Discussed over zoom, outcome: we changed the wording here.